### ISSUE TITLE
[iOS] Missing external ID on email channel after setting external id with auth hash

### ISSUE THREAD

**Issue by leethree** [OneSignal/react-native-onesignal#1173]
**Description:**

The external ID is set on push channel correctly but missing on email channel.

![image](https://user-images.githubusercontent.com/648142/108369082-1b10a780-71f3-11eb-93fb-61c10fb5d929.png)

**Environment**

OneSignal React-Native SDK: 4.0.4
React Native: 0.63.4

It happens on iOS. Not sure about Android.

**Steps to Reproduce Issue:**

Code to register email and external user ID:

```ts
export const identify = async (userId: string, email?: string) => {
  const { idHash, emailHash } = await fetchOneSignalHashes();
  if (email && emailHash) {
    OneSignal.setEmail(email, emailHash);
  }
  OneSignal.setExternalUserId(userId, idHash, results => {
    logger.debug('OneSignal: setting external user id', results);
  });
};
```

**Anything else:**

Got this alert when I enabled visual log:
![image](https://user-images.githubusercontent.com/648142/108372833-1f3ec400-71f7-11eb-9bed-9ad93f5169fe.png)

**Comment by tyang1** [comment]
Hi @leethree ,
Thank you for sharing the details and the code snippet.
Both the code and this warning indicated that the external user id has already been set. I see that the external_user_id is not shown on the the email record, which could be that the email record was created before the external_user_id was set the first time.
﻿
﻿Do you mind trying setting the email record first, and then reset the external_user_id by calling the [removeExternalUserId](https://documentation.onesignal.com/docs/sdk-reference#removeexternaluserid-method), followed by the setExternalUserId method afterwards ?
﻿
﻿Let us know if this helps resolving the warning, and enabling adding the external_user_id successfully,
﻿
﻿Thanks

**Comment by leethree** [comment]
I updated the code into this:

```ts
export const identify = async (userId: string, email?: string) => {
  const { idHash, emailHash } = await fetchOneSignalHashes();
  if (email && emailHash) {
    OneSignal.setEmail(email, emailHash);
  }
  OneSignal.removeExternalUserId(results => {
    logger.debug('OneSignal: removing external user id', results);
    OneSignal.setExternalUserId(userId, idHash, results => {
      logger.debug('OneSignal: setting external user id', results);
    });
  });
};
```

`removeExternalUserId` and `setExternalUserId` are both successful 

![image](https://user-images.githubusercontent.com/648142/108701825-5b786a00-7500-11eb-9986-e130ba438d01.png)

But I still get the same warning: `WARNING: Attempted to set external user id, but XXXXX is already set`.

**Comment by leethree** [comment]
I think the order of calling `setEmail` and `setExternalUserId` shouldn't matter. The SDK should handle it regardless. There might be a bug in iOS SDK to prevent this from working properly. The Android SDK seems fine in my testing.

Specifically, `shouldUpdateExternalUserId` in OneSignal-iOS-SDK looks suspicious to me.

https://github.com/OneSignal/OneSignal-iOS-SDK/blob/ee389b95a51399ec6317b9f28bba741aba606788/iOS_SDK/OneSignalSDK/Source/OneSignal.m#L2537-L2540

**Comment by rgomezp** [comment]
>There might be a bug in iOS SDK to prevent this from working properly. The Android SDK seems fine in my testing.

@emawby might be able to comment here.

Sirui could you please try putting the external id setter in the callback of the email setter? That way we can be sure that it won't fire until the email has been successfully set.

When testing, please uninstall the app completely before reinstalling and use a new, previously unused email address.

Cheers

**Comment by leethree** [comment]
I tried this:
> putting the external id setter in the callback of the email setter

On fresh start, strangely it failed when setting external id for email:

![21614094379_ pic](https://user-images.githubusercontent.com/648142/108867188-d7e87700-75ec-11eb-882a-fc2a108efa7b.jpg)

And I found this in the log:
```
Request <OSRequestUpdateExternalUserId: 0x600001b47030> fail result error Error Domain=OneSignalError Code=400 "(null)" UserInfo={returned={
    errors =     (
        "invalid identifier_auth_hash"
    );
}}
```

It appears that setting external ID is successful for push channel, but not for email channel.

On subsequent app starts, it does exactly the same behaviour as my previous comments. External ID is always missing from email channel.
